### PR TITLE
Fix foldArguments when passed target taking 0 arguments

### PR DESCRIPTION
**Pull request by mikezhang1234567890**
The MethodHandles.foldArguments method throws an exception if the foldPosition parameter passed is not in range of the target handle's argument count. This PR fixes erroneously throwing an exception when the argument count and foldPosition are both 0.

Signed-off-by: Mike Zhang <mike.h.zhang@ibm.com>

Fixes https://github.com/eclipse/openj9/issues/6655

### CODE REVIEW

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -2982,7 +2982,7 @@ private static final MethodHandle foldArgumentsCommon(MethodHandle handle, int f
 		final int argIndexCount = argumentIndices.length;
 		int[] passedInArgumentIndices = EMPTY_ARG_POSITIONS;
 		
-		if ((foldPosition < 0) || (foldPosition >= handleTypeParamCount)) {
+		if ((foldPosition < 0) || (foldPosition > handleTypeParamCount) || ((handleTypeParamCount != 0) && (foldPosition == handleTypeParamCount))) {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]:
```suggestion
		if ((foldPosition < 0) 
		|| (foldPosition > handleTypeParamCount) 
		/* Note, folding at handleTypeParamCount is only valid when the handle takes no arguments */
		|| ((foldPosition == handleTypeParamCount) && (foldPosition != 0)))
		{
```
Can this be reformatted similar to this?  It makes the code clearer to be able to see the three cases separately.

**Comment by mikezhang1234567890** [comment]
Sure, will add tests.

**Comment by andrewcraik** [comment]
FYI @liqunl just to make sure there isn't a similar issue in the JIT side of some of these implementations.

**Comment by liqunl** [comment]
This issue exists during MethodHandle creation where the JIT is not involved, so the JIT is good.

**Comment by DanHeidinga** [comment]
@mikezhang1234567890 when do you think the tests will be ready for this?

**Comment by mikezhang1234567890** [comment]
Should be ready by tomorrow.

**Comment by headius** [comment]
Do you have a way I could work around this until the fix is in general circulation?

**Comment by DanHeidinga** [comment]
Jenkins test sanity xlinux jdk8

**Comment by DanHeidinga** [comment]
Jenkins test sanity osx jdk12

**Comment by mikezhang1234567890** [comment]
@DanHeidinga added tests for the public interfaces of `foldArguments` for when target has 0 args. Can you review again please?

**Comment by DanHeidinga** [comment]
I've launched PR builds on this as it looks good to me.  Assuming those pass, I'll merge them

**Comment by mikezhang1234567890** [comment]
@headius I don't believe there is a workaround if you need to use a target handler with 0 arguments.

**Comment by headius** [comment]
I will endeavor to get OpenJ9 running our invokedynamic-heavy test suites for y'all. Maybe we can shake out any other issues.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -2982,7 +2982,11 @@ private static final MethodHandle foldArgumentsCommon(MethodHandle handle, int f
 		final int argIndexCount = argumentIndices.length;
 		int[] passedInArgumentIndices = EMPTY_ARG_POSITIONS;
 		
-		if ((foldPosition < 0) || (foldPosition >= handleTypeParamCount)) {
+		if ((foldPosition < 0)
+		|| (foldPosition > handleTypeParamCount)
+		/* folding at handleTypeParamCount is only valid when the handle takes no arguments */
+		|| ((foldPosition == handleTypeParamCount) && (handleTypeParamCount != 0) ))
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]:
```suggestion
		|| ((foldPosition == handleTypeParamCount) && (handleTypeParamCount != 0))
```

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 3]
```diff
@@ -2982,7 +2982,11 @@ private static final MethodHandle foldArgumentsCommon(MethodHandle handle, int f
...
+		{
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]:
```suggestion
		) {
```

**Pull request merged**